### ISSUE TITLE
Support unit suffixes and conversions

### ISSUE THREAD

**Issue by nikclayton-dfinity** [hashicorp/terraform#25612]
### Current Terraform Version

```
0.12.24
```

### Use-cases

Reduce the potential for error by making configuration values more explicit.

### Attempted Solutions

Multiplying values by multiples of 1,024 to get the desired number.

### Proposal

When specifying a numeric value allow suffixes to indicate the actual value, and a modifier to scale the value if necessary.

This was prompted by seeing:

```
resource "proxmox_vm_qemu" "..." {
  ...
  memory = 16 * 1024 # MiB
  ...
```

in a configuration.

I have no way of knowing (and it turns out the provider documentation is silent on this) whether desired value is 16MiB, or the unit of the field is MiB and the desired value is 16GiB.

We could expand the comment, but we can do better.

I propose:

1. Values should support typical suffixes (KB, KiB, MB, MiB, ...) that, if present, scale the value appropriately.

So

```
  memory = 16GiB
```

would be unambigious, and the value would be 17,179,869,184.

However, this only works if the unit for the value is bytes, it might not be, and Terraform can't control that, so

2. An optional extra clause, introduced with `as` or `in` keywords, followed by another suffix that scales the value back down. So if this provider entry unit is MiB I could write:

```
  memory = 16GiB in MiB
```

or

```
  memory = 16GiB as MiB
```

which would also be unambigious, and the value would be 16,384.

### References

Not that I've found.

**Comment by apparentlymart** [comment]
Hi @nikclayton-dfinity! Thanks for sharing this use-case.

Adding units to the language is an interesting idea. I expect if we were to do it we'd want to include it both in syntax _and_ in the type system, so that e.g. "data storage size" could be a separate type in the type system and thus avoid the potential pitfall you mentioned at the end of your comment of assigning a value in the wrong units, by making it a type check error.

At one point we had intended to have a convention that arguments accepting united numbers (or really, any "higher level" type beyond the physical types the language supports) should be named with a suffix to be clear about how they will be interpreted:

```
  memory_gib = 12 * 1024
  start_timestamp =  "2020-03-01T00:00:00Z"
  # ...
```

However, in practice this convention wasn't followed universally and it is no longer reasonable to say that this is a convention any user could rely on: there are numerous examples of arguments that are just single nouns with no such suffix, often because the provider developers are trying to stick to a competing convention of making the provider schema follow as closely as possible the underlying API schema.

Whether this justifies the additional complexity of adding an entirely new concept to the language I'm not sure, but I'd like to connect this with hashicorp/terraform-plugin-sdk#223, where the SDK team is thinking about a related concern of representing higher-level types in the plugin SDK. It seems like these use-cases are connected in that they are both around how to model higher-level interpretations of values than our storage-type-oriented type system currently captures.

Thanks again!

**Comment by nikclayton-dfinity** [comment]
I agree that types would definitely be desirable, and that in their absence a strong convention on suffixes would be helpful (and that that ship has probably sailed).

Although that still results in ambiguity. In your example:

```
memory_gib = 12 * 1024
```

that reads like it's going to specify 12TiB as the value for memory, which sounds unlikely.

However, I disagree that the proposal needs to be tied to type support in the language.

A couple of approaches spring to mind.

You could model symbols like `GiB`, `KB` etc as context dependent postfix operators.

When the argument is a number they multiply the number by the relevant amount so that `12 GiB` is syntactic sugar for `12 * 1073741824`.

When the argument is the keyword `in` or `as` they divide by that same constant, so `12 GiB in MiB` is syntactic sugar for `12 * 1073741824 / 1048576`.

Alternatively, and slightly more simply, you could just add them as named constants in the language. So `GiB` is the predefined constant 1073741824.

As a user I would have to be slightly more explicit:

```
memory_mib = 12 * GiB / MiB
```

for example, but I think it's more readable than what we have at the moment.

**Comment by apparentlymart** [comment]
Thanks for those additional ideas.

Right now it feels to me that in order to justify the additional language surface area this feature represents, it ought to be more than just syntactic sugar and should instead be something that allows Terraform to understand better the user's intent and change behavior accordingly.

I think it would be unfortunate if there were an argument named `memory` whose documentation defined it as being in gigabytes and the user was to write the following:

```
  memory = 8 GiB
```

...and then have the provider interpret it as 8 billion gigabytes. This seems like it would _hurt_ readability, by making a declaration seem like it means something different than what it actually means, rather than the current situation where a declaration is just _ambiguous_ about what it means unless you add a comment, or -- ideally! -- a situation where it means exactly what it seems to mean, because Terraform is able to understand that `memory` is _expecting_ gigabytes and thus no multiplication is needed (If we were starting fresh then in theory all providers could declare that `memory` is a quantity given in _bytes_ and then the syntactic sugar approach could be suitable, but it seems too late to change that now and our experience with the suffix convention doesn't give me confidence that providers would reliably follow the convention anyway.)

As you say, the suffix convention has also clearly failed and so I don't consider it a suitable solution for this use-case either: providers are not likely to accept either the breaking change of renaming their arguments or the complexity of supporting two arguments that do essentially the same thing just for this, or else presumably they would _already_ have followed this convention in the first place.

I guess I disagree that I'd interpret my example the way you stated:

```hcl
   memory_gib = 12 * 1024
```

I would read that as "12 * 1024 gigabytes", not "12 terrabytes", but all of these tradeoffs are objective so it's reasonable for us to interpret the same things in different ways. With that said, if `GiB` were just a different way to write a number as in your final example then that doesn't really feel much better to me than your original example which had a comment clarifying what unit the argument was being given in.

Broadly-speaking any new addition to the language has to provide enough value to justify the cost of new users having to learn it and the Terraform team having to maintain it. It feels to me like unit syntax is pretty low value unless Terraform can actually check it's being used correctly and make use of the additional information it provides. I'm unsure if united values are common enough in typical Terraform configurations that it would be justified even _with_ the type-checking benefits, but I would like to see us do some research into that before deciding.
